### PR TITLE
bitsandbytes: simplify 8bit dequantization

### PR DESCRIPTION
**Pull request by matthewdouglas**
# What does this PR do?

Simplifies the dequantization for bitsandbytes int8 weights. There is a similar PR open in [huggingface/peft#2245](https://github.com/huggingface/peft/pull/2245/files).

In the upcoming bitsandbytes release we will have an added API for for int8 dequantization. For backwards compatibility, a simplified (but functionally equivalent) dequantization operation is performed.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@SunMarc @BenjaminBossan

### CODE REVIEW

**Comment by BenjaminBossan** [comment]
LGTM, thanks. The failing test `tests/models/xlm_roberta_xl/test_modeling_xlm_roberta_xl.py::XLMRobertaXLModelTest::test_assisted_decoding_matches_greedy_search_1_same` appears to be unrelated.

File: src/transformers/integrations/bitsandbytes.py [hunk 1]
```diff
@@ -363,13 +363,14 @@ def dequantize_bnb_weight(weight: "torch.nn.Parameter", dtype: "torch.dtype", st
...
-    if state.CxB is None:
-        state.CxB, state.SB = bnb.functional.transform(weight.data, to_order=state.formatB)
-    out32, Sout32 = bnb.functional.igemmlt(im, state.CxB, Sim, state.SB)
-    return bnb.functional.mm_dequant(out32, Sout32, SCim, state.SCB, bias=None).t().to(dtype)
+    if hasattr(bnb.functional, "int8_vectorwise_dequant"):
+        # Use bitsandbytes API if available (requires v0.45.0+)
+        dequantized = bnb.functional.int8_vectorwise_dequant(weight.data, state.SCB)
+    else:
+        # Multiply by (scale/127) to dequantize.
+        dequantized = weight.data * state.SCB.view(-1, 1) * 7.874015718698502e-3
```
Comment (SunMarc) [inline src/transformers/integrations/bitsandbytes.py:371]: This line equivalent to what we had before ?

**Review by SunMarc — APPROVED** [review]
LGTM ! Left a question

File: src/transformers/integrations/bitsandbytes.py (hunk 1 above)
Comment (matthewdouglas) [reply src/transformers/integrations/bitsandbytes.py:371]: Yes. It is also the same way the selected 8bit weights are dequantized for fp16 compute in the LLM.int8() implementation: https://github.com/bitsandbytes-foundation/bitsandbytes/blob/7dca70049566b5b1c55cbd67e1cb191729a98152/bitsandbytes/autograd/_functions.py#L380

The method that this replaces was multiplying the quantized weights by a quantized identity matrix (int8 * int8 -> int32):
$W_{INT32} = I_{INT8}*W_{INT8} $

The dequantization feature `mm_dequant` was intended for dequantizing int32 matmul results, and as such performs $W_{FP16} = (W_{INT32} * (SC_{im} \bigotimes SCB)) / (127 * 127)$.

image

In this new version, I'm avoiding the cost of creating the identity matrix, quantizing it, and performing the matmul realizing int32 values. Mathematically this is equivalent, and more importantly it's simpler and introduces less error.

**Review by ArthurZucker — APPROVED** [review]
A welcome change! THanks 🤗

File: src/transformers/integrations/bitsandbytes.py (hunk 1 above)
Comment (SunMarc) [reply src/transformers/integrations/bitsandbytes.py:371]: Nice !

**Pull request merged**